upup/pkg/fi/clientset_castore.go

```diff
@@ -0,0 +1,630 @@
+	for _, ki := range keyset.Spec.Keys {
+		if ki.Id == id {
+			found = true
+		} else {
+			newKeys = append(newKeys, ki)
+		}
+	}
+	if !found {
+		return fmt.Errorf("KeysetItem %q not found in Keyset %q", id, name)
+	}
+	if len(newKeys) == 0 {
+		if err := client.Delete(name, &v1.DeleteOptions{}); err != nil {
+			return fmt.Errorf("error deleting Keyset %q: %v", name, err)
+		}
+	} else {
+		keyset.Spec.Keys = newKeys
+		if _, err := client.Update(keyset); err != nil {
+			return fmt.Errorf("error updating Keyset %q: %v", name, err)
+		}
+	}
+	return nil
+}
+
+// addSshCredential saves the specified SSH Credential to the registry, doing an update or insert
+func (c *ClientsetCAStore) addSshCredential(name string, publicKey string) error {
+	create := false
+	client := c.clientset.SSHCredentials(c.namespace)
+	sshCredential, err := client.Get(name, v1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			sshCredential = nil
+		} else {
+			return fmt.Errorf("error reading SSHCredential %q: %v", name, err)
+		}
+	}
+	if sshCredential == nil {
+		sshCredential = &kops.SSHCredential{}
+		sshCredential.Name = name
+		create = true
+	}
+	sshCredential.Spec.PublicKey = publicKey
+	if create {
+		if _, err := client.Create(sshCredential); err != nil {
+			return fmt.Errorf("error creating SSHCredential %q: %v", name, err)
+		}
+	} else {
+		if _, err := client.Update(sshCredential); err != nil {
+			return fmt.Errorf("error updating SSHCredential %q: %v", name, err)
+		}
+	}
+	return nil
+}
+
+// deleteSSHCredential deletes the specified SSHCredential from the registry
+func (c *ClientsetCAStore) deleteSSHCredential(name string) error {
+	client := c.clientset.SSHCredentials(c.namespace)
+	err := client.Delete(name, &v1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("error deleting SSHCredential %q: %v", name, err)
+	}
+	return nil
+}
+
+// addKey saves the specified keypair to the registry
+func (c *ClientsetCAStore) storeKeypair(name string, id string, cert *pki.Certificate, privateKey *pki.PrivateKey) error {
+	var publicMaterial bytes.Buffer
+	if _, err := cert.WriteTo(&publicMaterial); err != nil {
+		return err
+	}
+
+	var privateMaterial bytes.Buffer
+	if _, err := privateKey.WriteTo(&privateMaterial); err != nil {
+		return err
+	}
+
+	item := &kops.KeysetItem{
+		Id:              id,
+		PublicMaterial:  publicMaterial.Bytes(),
+		PrivateMaterial: privateMaterial.Bytes(),
+	}
+	return c.addKey(name, kops.SecretTypeKeypair, item)
+}
+
+// buildSerial returns a serial for use when issuing certificates
+func (c *ClientsetCAStore) buildSerial() *big.Int {
+	t := time.Now().UnixNano()
+	return pki.BuildPKISerial(t)
+}
+
+// AddSSHPublicKey implements CAStore::AddSSHPublicKey
+func (c *ClientsetCAStore) AddSSHPublicKey(name string, pubkey []byte) error {
+	_, _, _, _, err := ssh.ParseAuthorizedKey(pubkey)
+	if err != nil {
+		return fmt.Errorf("error parsing SSH public key: %v", err)
+	}
+
+	// TODO: Reintroduce or remove
```

Comment:
What are we loosing by not having this?

Reply:
Unclear at the present time... we have nowhere to store it on the SSHCredential API type (previously they were on the Keyset API type)